californium-core/src/main/java/org/eclipse/californium/core/network/UdpMatcher.java

```diff
@@ -310,11 +317,19 @@ public void completed(final Exchange exchange) {
 				} else {
 					// in case an empty ACK was lost
 					KeyMID idByMID = KeyMID.fromOutboundMessage(originRequest);
-					exchangeStore.remove(idByMID);
+					exchangeStore.remove(idByMID, exchange);
 				}
 				KeyToken idByToken = KeyToken.fromOutboundMessage(originRequest);
-				exchangeStore.remove(idByToken);
-				if(!originRequest.getOptions().hasObserve()) {
+				exchangeStore.remove(idByToken, exchange);
+				Request request = exchange.getRequest();
+				if (request != originRequest && null != request &&  null != request.getToken()
+						&& !Arrays.equals(request.getToken(), originRequest.getToken())) {
+					// remove starting request also
+					originRequest = request;
+					idByToken = KeyToken.fromOutboundMessage(originRequest);
+					exchangeStore.remove(idByToken, exchange);
```

Comment:
@sbernard31 

I checked now the possibility to move the blockwise specific exchange stuff into the "blockwise layer".

Result:
Exchange only offers 1 ExchangeObserver, it could only be set not get
The UdpMatcher.ExchangeObserverImpl is private and could therefore not be extended.

So I don't see a way to move the stuff without making too complicated.